File: LeetCode/Medium/Q54.java
package LeetCode.Medium;

import java.util.ArrayList;
import java.util.List;

public class Q54 {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> mat = new ArrayList<>();
        int startRow = 0;
        int startCol = 0;
        int endRow = matrix.length - 1;
        int endCol = matrix[0].length - 1;

        while (startRow <= endRow && startCol <= endCol) {
            for (int i = startCol; i <= endCol; i++) {
                mat.add(matrix[startRow][i]);
            }
            startRow++;

            for (int i = startRow; i <= endRow; i++) {
                mat.add(matrix[i][endCol]);
            }
            endCol--;

            if (startRow <= endRow) {
                for (int i = endCol; i >= startCol; i--) {
                    mat.add(matrix[endRow][i]);
                }
                endRow--;
            }

            if (startCol <= endCol) {
                for (int i = endRow; i >= startRow; i--) {
                    mat.add(matrix[i][startCol]);
                }
                startCol++;
            }
        }
        return mat;
    }
}

File: LeetCode/Python/Medium/Q54.py
class Solution:
    def spiralOrder(self, matrix):
        m, n = len(matrix), len(matrix[0])
        mat = []
        rStart, rEnd = 0, m - 1
        cStart, cEnd = 0, n - 1
        while rStart <= rEnd and cStart <= cEnd:
            for i in range(cStart, cEnd + 1):
                mat.append(matrix[rStart][i])
            rStart += 1
            for i in range(rStart, rEnd + 1):
                mat.append(matrix[i][cEnd])
            cEnd -= 1
            if rStart <= rEnd:
                for i in range(cEnd, cStart - 1, -1):
                    mat.append(matrix[rEnd][i])
                rEnd -= 1
            if cStart <= cEnd:
                for i in range(rEnd, rStart - 1, -1):
                    mat.append(matrix[i][cStart])
                cStart += 1
        return mat
